File: CarHomeWidget/src/com/anod/car/home/appwidget/ShortcutPendingIntent.java
package com.anod.car.home.appwidget;

import android.app.PendingIntent;
import android.appwidget.AppWidgetManager;
import android.content.Context;
import android.content.Intent;
import android.net.Uri;

import com.anod.car.home.ShortcutActivity;
import com.anod.car.home.incar.ModeService;
import com.anod.car.home.prefs.Configuration;
import com.anod.car.home.prefs.PickShortcutUtils;
import com.anod.car.home.utils.IntentUtils;

public class ShortcutPendingIntent implements LauncherViewBuilder.PendingIntentHelper {
	public static final String INTENT_ACTION_CALL_PRIVILEGED = "android.intent.action.CALL_PRIVILEGED";
	final private Context mContext;

	public ShortcutPendingIntent(Context context) {
		mContext = context;
	}

	/**
     * Create an Intent to launch Configuration
     * @param appWidgetId
     * @return
     */
    @Override
    public PendingIntent createSettings(int appWidgetId, int cellId) {
		Intent intent = IntentUtils.createSettingsIntent(mContext, appWidgetId, cellId);
    	return PendingIntent.getActivity(mContext, 0, intent, PendingIntent.FLAG_UPDATE_CURRENT);
    }

	@Override
    public PendingIntent createShortcut(Intent intent,int appWidgetId, int position, long shortcutId) {
    	return createShortcut(intent,String.valueOf(appWidgetId),position);
    }
    
    public PendingIntent createShortcut(Intent intent,String prefix, int position) {
    	String action = intent.getAction();
    	boolean isCallPrivileged = (action != null && action.equals(INTENT_ACTION_CALL_PRIVILEGED));
    	if (intent.getExtras() == null && !isCallPrivileged) { // Samsung s3 bug
    		return PendingIntent.getActivity(mContext, 0 /* no requestCode */, intent, PendingIntent.FLAG_UPDATE_CURRENT);
    	}
    	String path = prefix + " - " + position;
    	Uri data = Uri.withAppendedPath(Uri.parse("com.anod.car.home://widget/id/"),path);
    	
    	if (action != null && action.equals(ShortcutActivity.ACTION_MEDIA_BUTTON)) {
    		intent.setData(data);
    		return PendingIntent.getActivity(mContext, 0 /* no requestCode */, intent, PendingIntent.FLAG_UPDATE_CURRENT);
    	}
    	
    	Intent launchIntent = new Intent(mContext, ShortcutActivity.class);
    	launchIntent.setData(data);
    	launchIntent.setAction(Intent.ACTION_MAIN);
    	launchIntent.putExtra(ShortcutActivity.EXTRA_INTENT, intent);
		return PendingIntent.getActivity(mContext, 0 /* no requestCode */, launchIntent, PendingIntent.FLAG_UPDATE_CURRENT);
    }

	@Override
	public PendingIntent createInCar(boolean on) {
		if (on) {
			return getInCarOnIntent();
		}
		return getInCarOffIntent();
	}

	private PendingIntent getInCarOnIntent() {
		Intent onIntent = new Intent(mContext, ModeService.class);
		onIntent.putExtra(ModeService.EXTRA_MODE, ModeService.MODE_SWITCH_ON);
		onIntent.putExtra(ModeService.EXTRA_FORCE_STATE, true);
		Uri data = Uri.parse("com.anod.car.home.pro://mode/1/1");
		onIntent.setData(data);
		return PendingIntent.getService(mContext, 0, onIntent, 0);
	}

	private PendingIntent getInCarOffIntent() {
		Intent offIntent = new Intent(mContext, ModeService.class);
		offIntent.putExtra(ModeService.EXTRA_MODE, ModeService.MODE_SWITCH_OFF);
		offIntent.putExtra(ModeService.EXTRA_FORCE_STATE, true);
		Uri data = Uri.parse("com.anod.car.home.pro://mode/0/1");
		offIntent.setData(data);
		return PendingIntent.getService(mContext, 0, offIntent, 0);
	}
}

File: app/src/main/java/com/anod/car/home/appwidget/ShortcutPendingIntent.kt
package com.anod.car.home.appwidget

import com.anod.car.home.ShortcutActivity
import com.anod.car.home.incar.ModeService

import android.app.PendingIntent
import android.appwidget.AppWidgetManager
import android.content.Context
import android.content.Intent
import android.net.Uri
import com.anod.car.home.OverlayActivity
import info.anodsplace.carwidget.appwidget.PendingIntentFactory
import info.anodsplace.carwidget.content.Deeplink
import info.anodsplace.carwidget.content.shortcuts.ShortcutExtra
import info.anodsplace.carwidget.content.shortcuts.ShortcutResources
import info.anodsplace.carwidget.utils.forSettings

class ShortcutPendingIntent(private val context: Context, private val shortcutResources: ShortcutResources) : PendingIntentFactory {

    private val inCarOnIntent: PendingIntent
        get() {
            val onIntent = Intent(context, ModeService::class.java)
            onIntent.putExtra(ModeService.EXTRA_MODE, ModeService.MODE_SWITCH_ON)
            onIntent.putExtra(ModeService.EXTRA_FORCE_STATE, true)
            onIntent.data = Deeplink.SwitchMode(true).toUri()
            return PendingIntent.getService(context, 0, onIntent, PendingIntent.FLAG_IMMUTABLE)
        }

    private val inCarOffIntent: PendingIntent
        get() {
            val offIntent = Intent(context, ModeService::class.java)
            offIntent.putExtra(ModeService.EXTRA_MODE, ModeService.MODE_SWITCH_OFF)
            offIntent.putExtra(ModeService.EXTRA_FORCE_STATE, true)
            offIntent.data = Deeplink.SwitchMode(false).toUri()
            return PendingIntent.getService(context, 0, offIntent, PendingIntent.FLAG_IMMUTABLE)
        }

    override fun createNew(appWidgetId: Int, position: Int): PendingIntent {
        val editIntent = Intent(
                Intent.ACTION_VIEW,
                Deeplink.EditShortcut(appWidgetId, 0, position).toUri(),
                context,
                OverlayActivity::class.java
        ).also {
            it.putExtra(AppWidgetManager.EXTRA_APPWIDGET_ID, appWidgetId)
        }

        return PendingIntent.getActivity(context, 0, editIntent, PendingIntent.FLAG_UPDATE_CURRENT or PendingIntent.FLAG_IMMUTABLE)
    }

    /**
     * Create an Intent to launch Configuration
     */
    override fun createSettings(appWidgetId: Int, buttonId: Int): PendingIntent {
        val intent = Intent().forSettings(context, appWidgetId, shortcutResources)
        return PendingIntent.getActivity(context, 0, intent, PendingIntent.FLAG_UPDATE_CURRENT or PendingIntent.FLAG_IMMUTABLE)
    }

    override fun createShortcut(intent: Intent, appWidgetId: Int, position: Int, shortcutId: Long): PendingIntent? {
        return createShortcut(intent) { Deeplink.OpenWidgetShortcut(appWidgetId, position).toUri() }
    }

    fun createShortcut(intent: Intent, uri: () -> Uri): PendingIntent? {
        val action = intent.action ?: ""
        val isCall = INTENT_ACTION_CALL_PRIVILEGED == action || Intent.ACTION_CALL == action
        if (intent.extras == null && !isCall) { // Samsung s3 bug
            return PendingIntent.getActivity(context, 0, intent, PendingIntent.FLAG_UPDATE_CURRENT or PendingIntent.FLAG_IMMUTABLE)
        }

        val data = uri()

        if (action == ShortcutExtra.ACTION_MEDIA_BUTTON) {
            intent.data = data
            return PendingIntent.getActivity(context, 0, intent, PendingIntent.FLAG_UPDATE_CURRENT or PendingIntent.FLAG_IMMUTABLE)
        }

        val launchIntent = Intent(context, ShortcutActivity::class.java)
        launchIntent.flags = Intent.FLAG_ACTIVITY_NEW_TASK or Intent.FLAG_ACTIVITY_NO_HISTORY
        launchIntent.data = data
        launchIntent.action = Intent.ACTION_MAIN
        launchIntent.putExtra(ShortcutActivity.EXTRA_INTENT, intent)
        return PendingIntent.getActivity(context, 0, launchIntent, PendingIntent.FLAG_UPDATE_CURRENT or PendingIntent.FLAG_IMMUTABLE)
    }

    override fun createInCar(on: Boolean, buttonId: Int): PendingIntent {
        return if (on) {
            inCarOnIntent
        } else inCarOffIntent
    }

    companion object {
        const val INTENT_ACTION_CALL_PRIVILEGED = "android.intent.action.CALL_PRIVILEGED"
    }
}
